tiny_dnn/core/kernels/global_avepool_op_avx.h

```diff
@@ -0,0 +1,186 @@
+/*
+    Copyright (c) 2017, Taiga Nomi
+    All rights reserved.
+
+    Use of this source code is governed by a BSD-style license that can be found
+    in the LICENSE file.
+*/
+#pragma once
+
+namespace tiny_dnn {
+namespace kernels {
+
+#ifdef CNN_USE_AVX
+#ifdef CNN_USE_DOUBLE
+
+// double version
+inline void global_avepool_op_avx(const tensor_t &in_data,
+                                  tensor_t &out_data,
+                                  const core::global_avepool_params &params,
+                                  const bool layer_parallelize) {
+  const size_t pool_area            = params.in.width_ * params.in.height_;
+  const double pool_area_inv        = 1.0f / static_cast<double>(pool_area);
+  const size_t nblocks_per_channel  = pool_area / 4;
+  const size_t nremains_per_channel = pool_area & 3;
+
+  int64_t mask_src[] = {-1, -1, -1, -1, 0, 0, 0, 0};
+  __m256i imask =
+    _mm256_loadu_si256((__m256i const *)(mask_src + 4 - nremains_per_channel));
+
+  for_i(layer_parallelize, in_data.size(), [&](size_t sample) {
+    const vec_t &in = in_data[sample];
+    vec_t &out      = out_data[sample];
+
+    for (size_t i = 0; i < params.in.depth_; i++) {
+      __m256d sum_m            = _mm256_setzero_pd();
+      const size_t depth_index = i * pool_area;
+      for (size_t j = 0; j < nblocks_per_channel; j++) {
+        __m256d in_m = _mm256_loadu_pd(&in[depth_index + 4 * j]);
+        sum_m        = _mm256_add_pd(sum_m, in_m);
```

Comment:
@karandesai-96 
> What if nblocks_per_channel is less than 8 ?

You can process remaining data with not unrolled code.
Having 4x and 2x loop unrolling is also fine.

I recommend you to add `vectorize::accumulate` in `tiny_dnn/util/product.h` file.
In that way, reusing optimized code would be easy.